warehouse/manage/views.py

```diff
@@ -1059,6 +1070,232 @@ def toggle_2fa_requirement(self):
+                resets_in=self._ratelimiters["ip.oidc"].resets_in(self.remote_addr)
+            )
+
+    @property
+    def github_provider_form(self):
+        return GitHubProviderForm(
+            self.request.POST,
+            api_token=self.request.registry.settings.get("github.token"),
+        )
+
+    @property
+    def default_response(self):
+        return {
+            "project": self.project,
+            "github_provider_form": self.github_provider_form,
+        }
+
+    @view_config(request_method="GET")
+    def manage_project_oidc_providers(self):
+        if self.request.flags.enabled(AdminFlagValue.DISALLOW_OIDC):
+            self.request.session.flash(
+                (
+                    "OpenID Connect is temporarily disabled. "
+                    "See https://pypi.org/help#admin-intervention for details."
+                ),
+                queue="error",
+            )
+
+        return self.default_response
+
+    @view_config(request_method="POST", request_param=GitHubProviderForm.__params__)
+    def add_github_oidc_provider(self):
+        if self.request.flags.enabled(AdminFlagValue.DISALLOW_OIDC):
+            self.request.session.flash(
+                (
+                    "OpenID Connect is temporarily disabled. "
+                    "See https://pypi.org/help#admin-intervention for details."
+                ),
+                queue="error",
+            )
+            return self.default_response
+
+        try:
+            self._check_ratelimits()
+        except TooManyOIDCRegistrations as exc:
+            return HTTPTooManyRequests(
+                self.request._(
+                    "There have been too many attempted OpenID Connect registrations. "
+                    "Try again later."
+                ),
+                retry_after=exc.resets_in.total_seconds(),
+            )
+
+        response = self.default_response
+        form = response["github_provider_form"]
+
+        if form.validate():
+            self._hit_ratelimits()
+            # GitHub OIDC providers are unique on the tuple of
+            # (repository_name, owner, workflow_name), so we check for
+            # an already registered one before creating.
+            provider = (
+                self.request.db.query(GitHubProvider)
+                .filter(
+                    GitHubProvider.repository_name == form.repository.data,
+                    GitHubProvider.owner == form.normalized_owner,
+                    GitHubProvider.workflow_name == form.workflow_name.data,
+                )
+                .one_or_none()
+            )
+            if provider is None:
+                provider = GitHubProvider(
+                    repository_name=form.repository.data,
+                    owner=form.normalized_owner,
+                    owner_id=form.owner_id,
+                    workflow_name=form.workflow_name.data,
+                )
+
+                self.request.db.add(provider)
```

Comment:
This has the potential effect of adding a duplicate provider. We need to check if the provider exists before adding it.

Reply:
I might be misunderstanding what you mean, but isn't that what the `one_or_none()` query above does? We only create the provider if it doesn't already exist.

Reply:
Whoops, didn't mean to resolve.